---
title: "explore.taxonomy"
author: "M Fisher"
date: "1/21/2021"
output: 
  html_document:
    toc: yes
    toc_float: yes
---

# Description

Explorative plots looking at taxonomic assignments from **BF3** sequences.


# Set up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
<br>

This script requires the following packages:
```{r message=FALSE, warning=FALSE}
library(here)
library(tidyverse)
library(magrittr)
library(cowplot)
```
<br>

This script calls the following custom function:
```{r}
source(here('R','custom_lca.R'))
```
<br>

User directories
```{r}
indir_blast  <- 'test-lane-1/data/blast'
indir_asv    <- 'test-lane-1/data/dada2/BF3'
indir_meta   <- 'test-lane-1/data/cutadapt/BF3'
outdir       <- 'test-lane-1/results'
```
<br>

Which marker is being analyzed?
```{r}
marker <- 'BF3'
```
<br>

What are the names of the processed blast file with taxonomy from script 6?
```{r}
# hashes with unique staxids
infile1 <- 'lane1_bf3_blast_uniqueIDs_taxonomy.csv'

# hashes with multiple IDs, that required the LCA function
infile2 <- 'lane1_bf3_blast_multipleIDs_taxonomy.csv'
```
<br>
<br>


# Data

Taxonomy from script `6.parse.blast.results`
```{r}
tax_key <- read_csv(here(indir_blast, infile2)) %>%
  rename("Hash"=representative)

tax_key %<>% bind_rows(read_csv(here(indir_blast, infile1)) %>%
              rename("Hash"=qseqid) %>%
                mutate(taxID="unique", score="unique") %>%
              dplyr::select(all_of(colnames(tax_key))))
```
<br>

ASV table
```{r}
dat <- read_csv(here(indir_asv, 'merged_ASV_table.csv'))
```
<br>

Sequencing sample metadata, so that we can connect sample name (actually a number; in the ASV table) to sample ID.
```{r}
metadat <- read_csv(here(indir_meta, paste0('cutadapt_output_metadata_',marker,'.csv')))

metadat %<>% dplyr::select(Sample_name, file1) %>%
  mutate(SampleID=str_remove(file1,"_L001_R1_001.fastq.fastq")) %>%
  mutate(SampleID=str_remove(SampleID,paste0("Locus_",marker,"_",marker,"_"))) %>%
  mutate(SampleID=str_sub(SampleID,start=1, end=-5)) %>%
  mutate(Group=ifelse(SampleID %in% c("Kangaroo","PCRNegative","SpeedVacNegative"), "Control", str_sub(SampleID, start=1,end=4))) %>%
  dplyr::select(-file1)
```
<br>

Join the sample metadata to the ASV dataframe
```{r}
dat %<>% left_join(metadat,by="Sample_name")
```
<br>

Join the taxonomic key to the ASV dataframe
```{r}
dat %<>% left_join(tax_key,by="Hash")
```
<br>

### quality check

**1.** Make sure that each hash has only one set of taxonomic information
```{r}
checkdat <- dat %>% 
  group_by(Sample_name,Hash) %>%
  summarise(count=n(), .groups="drop") %>%
  filter(count > 1)
checkdat
```
<br>

It looks like all of those with a '2' count have matching taxonomy between the two unique IDs, but one row of taxonomic info is more fine-scale than the other. Save only the most specific taxonomic info.
```{r}
## rows to remove have the most NAs
checkdat %<>% left_join(dat,by=c("Sample_name","Hash")) %>%
  arrange(desc(rowSums(is.na(.)))) %>%
  distinct(Sample_name, .keep_all = TRUE)

## remove those rows
dat %<>% anti_join(checkdat, by=c("Sample_name","Hash", "rank", "taxon"))

rm(checkdat)
```
<br>

```{r}
dat %>% 
  group_by(Sample_name,Hash) %>%
  summarise(count=n(), .groups="drop") %>%
  filter(count > 1)
```
<br>

**2.** Find the number of hashes with missing taxonomic information
```{r fig.width=10}
## total hashes / reads
hcount <- dat %>%
  group_by(Sample_name,Group,SampleID) %>%
  summarise(total_hashes=n(),
            total_reads=sum(nReads),.groups="drop")

## proportion of hashes / reads missing taxonomy
checkdat <- dat %>%
  filter(is.na(rank)) %>%
  group_by(Sample_name,Group,SampleID) %>%
  summarise(nhash=n(),
            nreads=sum(nReads),.groups="drop") %>%
  left_join(hcount,by=c("Sample_name","Group","SampleID")) %>%
  mutate(hashes_na=nhash/total_hashes,
         reads_na=nreads/total_reads) %>%
    pivot_longer(cols=c(hashes_na, reads_na), names_to="metric")

ggplot(checkdat, aes(x=as.character(SampleID),y=value, fill=Group)) +
  geom_col() +
  facet_grid(col=vars(Group), row=vars(metric), scales="free") +
  scale_y_continuous(limits=c(0,1)) +
  xlab("Sample ID") + ylab("Proportion w/o Taxonomy") +
  theme_bw() + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),
                     legend.position="none")
```
<br>
<br>

# Ratios

A quick look at the proportion of reads belonging to each taxa, for every sample. 

## Controls

What DNA did we pick up in the controls?

```{r}
control_dat <- dat %>% filter(Group=="Control" & dataset=="merged")
```
<br>

```{r echo=FALSE}
ggplot(control_dat, aes(x=SampleID, y=nReads, fill=phylum)) +
  geom_col() +
  theme_bw() + ggtitle("Phyla")
```
<br>

```{r echo=FALSE}
ggplot(control_dat, aes(x=SampleID, y=nReads, fill=taxon)) +
  geom_col() +
  theme_bw() + ggtitle("Taxa")
```
<br>

It looks like the green crab DNA got...everywhere. The rest is mostly bacteria. The weirdest thing is that there isn't any kangaroo DNA in the positive control??

<br>

## Predator v Prey

How much predator DNA did we pick up, from each sample? 

### Green Crab
```{r}
egc_dat <- dat %>%
  filter(Group %in% c("WACO","WASS")) %>%
  filter(dataset=="merged") %>%
  mutate(Predator=ifelse(taxon == "Carcinus maenas" | taxon == "Carcinus aestuarii",1,0))
```
<br>

```{r echo=FALSE}
ggplot(egc_dat, aes(x=SampleID, y=nReads, fill=as.factor(Predator))) +
  geom_col() +
  facet_grid(col=vars(Group), scales="free") +
  scale_fill_manual(values=c("lightblue","red"), name="Predator") +
  theme_bw() + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
```
<br>


### Dungies
```{r}
dungy_dat <- dat %>%
  filter(Group=="MARP") %>%
  filter(dataset=="merged") %>%
  mutate(Predator=ifelse(taxon == "Metacarcinus magister",1,0))
```
<br>

```{r echo=FALSE}
ggplot(dungy_dat, aes(x=SampleID, y=nReads, fill=as.factor(Predator))) +
  geom_col() +
  scale_fill_manual(values=c("lightblue","red"), name="Predator") +
  theme_bw() + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
```
<br>

## Diet Composition

### Green Crab

```{r}
egc_prey <- egc_dat %>%
  filter(Predator==0)
```
```{r echo=FALSE}
ggplot(egc_prey, aes(x=SampleID, y=nReads, fill=as.factor(phylum))) +
  geom_col() + facet_grid(cols=vars(Group), scales="free") +
  ggtitle("Phyla") +
  theme_bw() + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
```
<br>

I suspect the *Chordata* is contamination from me or from the SEFS lab...
```{r echo=FALSE}
egc_prey %>% filter(phylum=="Chordata") %>% dplyr::select(SampleID,taxon) %>% distinct()
```
<br>

Weird, it almost looks like WACO_2021_001a should be the positive control???

The wolf DNA in WACO_2021_002c is contamination from SEFS. Let's remove the *Chordata* DNA for now. 
```{r}
egc_prey %<>% filter(phylum!="Chordata")
```
```{r echo=FALSE}
ggplot(data=filter(egc_prey, phylum %in% c("Arthropoda","Mollusca","Cnidaria")), aes(x=SampleID, y=nReads, fill=taxon)) +
  geom_col() + facet_grid(cols=vars(Group), scales="free") +
  ggtitle("Arthropoda, Mollusca, & Cnidaria") +
  theme_bw() + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
```
<br>

So we have lots of Dungeness crab and *Hemigrapsus* reads, with some *Ampithoe lacertosa*, an amphipod. The *Mya spp.* are soft-shell clams.

```{r echo=FALSE}
ggplot(data=filter(egc_prey, phylum %in% c("Bacillariophyta","Chlorophyta")), aes(x=SampleID, y=nReads, fill=taxon)) +
  geom_col() + facet_grid(cols=vars(Group), scales="free") +
  ggtitle("Algae & Diatoms") +
  theme_bw() + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
```
<br>

## Dungies

```{r}
dungy_prey <- dungy_dat %>%
  filter(Predator==0)
```
```{r echo=FALSE}
ggplot(dungy_prey, aes(x=SampleID, y=nReads, fill=phylum)) +
  geom_col() +
  ggtitle("Phyla") +
  theme_bw() + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
```
<br>

Not a lot of diversity in these samples.

```{r echo=FALSE}
ggplot(data=dungy_prey, aes(x=SampleID, y=nReads, fill=taxon)) +
  geom_col() +
  ggtitle("Taxa") +
  theme_bw() + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))
```
<br>

Well, crap, it looks like there's just a bunch of contamination from the green crab. Plus some brown algae.

<br>
<br>


# eDNA Index











