---
title: "Running BLAST shell (notes)"
author: "M Fisher (via Eily via Moncho)"
date: '2022-09-23'
output: html_document
---

# Description 

These notes describe the bash shell script used to match the ASV DNA sequences from DADA2 to NCBI's nucleotide database. I run everything in an **Ubuntu for Windows** terminal, not through R. 

**Ubuntu for Windows**: Ubuntu 22.04 LTS app, run in Windows 10

*Dependencies*: wget, gunzip (downloaded using `sudo apt install`)



# Installing / Updating BLAST

Install the blast database - see [Install BLAST](https://github.com/mfisher5/EGC-Willapa-Bay/blob/main/doc/Install_BLAST.txt) in the doc/ folder

Update the blast database
```{bash eval=FALSE}
cd /mnt/c/Users/mfisher5/Documents/my_programs/ncbi-blast-2.13.0+/blastdb
../bin/update_blastdb.pl --decompress nt
```
<br>

make sure that the program `libgomp1` is installed
```{bash eval=FALSE}
sudo apt-get install libgomp1
```
<br>

# Running BLAST

## sept 23

The command in the terminal will be: 
```
cd /mnt/c/Users/mfisher5/Documents/my_programs/ncbi-blast-2.13.0+/bin/
bash /mnt/c/Users/mfisher5/Documents/EGC-Salish-Sea/scripts/3_run_blast_run2.sh
```

```{bash eval=FALSE}
#!/usr/bin/env bash
PATH=$PATH:/mnt/c/Users/mfisher5/Documents/my_programs/ncbi-blast-2.13.0+/bin
export BLASTDB="/mnt/c/Users/mfisher5/Documents/my_programs/ncbi-blast-2.13.0+/blastdb"
BLAST_DB='/mnt/c/Users/mfisher5/Documents/my_programs/ncbi-blast-2.13.0+/blastdb/nt'
blast_output='/mnt/c/Users/mfisher5/Documents/EGC-Salish-Sea/data/blast/hash_key_blast_2022-09-23.fasta'
# BLAST PARAMETERS
PERCENT_IDENTITY="85"
WORD_SIZE="30"
EVALUE="1e-30"
# number of matches recorded in the alignment:
MAXIMUM_MATCHES="50"
CULLING="5"


	################################################################################
	# BLAST CLUSTERS
	################################################################################
	echo $(date +%H:%M) "BLASTing..." 
	./blastn \
		-query "/mnt/c/Users/mfisher5/Documents/EGC-Salish-Sea/data/dada2/hash_key.fasta
" \
		-db "${BLAST_DB}" \
		-num_threads 16 \
		-perc_identity "${PERCENT_IDENTITY}" \
		-word_size "${WORD_SIZE}" \
		-evalue "${EVALUE}" \
		-max_target_seqs "${MAXIMUM_MATCHES}" \
		-culling_limit="${CULLING}" \
		-outfmt "6 qseqid sseqid sacc pident length mismatch gapopen qcovus qstart qend sstart send evalue bitscore staxids qlen sscinames sseq" \
		-out "${blast_output}"
```

## oct 25

I'm going to try running blast with a lower word size. In the sept blast output, I ended up with a hit for a non-native species that hasn't been recorded in the Puget Sound (ascophyllum nodosum). When I manually blasted the sequence that returned this ID, the output showed *Hormosira banksii* at 87% (also non native), and *Fucus spiralis* (84.5%). Yet Ascophyllum is what is being written out as the only ID for that sequence.

Ascophyllum nodosum (knotted wrack): 2 crab from Drayton Harbor (85.2% identity in blast, one ASV)

Kurtiella bidentata (small bivalve): 3 crab from Drayton Harbor (>99% identity in blast, three ASVs)

This iteration of blast, I'm going to change the `word size` parameter to 15. From Moncho: "For what I looked up, that is the size of a perfect match for which to start growing the alignment score. If there is nothing too closely related on your db, a long word size will reduce the number of hits to begin with. And with COI, in which every third base can freely change, that is maybe reducing the number of leads. The default value of word size for blastn is 11."

The command in the terminal will be: 
```
cd /mnt/c/Users/mfisher5/Documents/my_programs/ncbi-blast-2.13.0+/bin/
bash /mnt/c/Users/mfisher5/Documents/EGC-Salish-Sea/scripts/testing_parameters/3_run_blast_run2-ws15.sh
```

```{bash eval=FALSE}
#!/usr/bin/env bash
PATH=$PATH:/mnt/c/Users/mfisher5/Documents/my_programs/ncbi-blast-2.13.0+/bin
export BLASTDB="/mnt/c/Users/mfisher5/Documents/my_programs/ncbi-blast-2.13.0+/blastdb"
BLAST_DB='/mnt/c/Users/mfisher5/Documents/my_programs/ncbi-blast-2.13.0+/blastdb/nt'
blast_output='/mnt/c/Users/mfisher5/Documents/EGC-Salish-Sea/data/blast/hash_key_blast_2022-10-25-ws15.fasta'
# BLAST PARAMETERS
PERCENT_IDENTITY="85"
WORD_SIZE="30"
EVALUE="1e-30"
# number of matches recorded in the alignment:
MAXIMUM_MATCHES="15"
CULLING="5"


	################################################################################
	# BLAST CLUSTERS
	################################################################################
	echo $(date +%H:%M) "BLASTing..." 
	./blastn \
		-query "/mnt/c/Users/mfisher5/Documents/EGC-Salish-Sea/data/dada2/hash_key.fasta
" \
		-db "${BLAST_DB}" \
		-num_threads 16 \
		-perc_identity "${PERCENT_IDENTITY}" \
		-word_size "${WORD_SIZE}" \
		-evalue "${EVALUE}" \
		-max_target_seqs "${MAXIMUM_MATCHES}" \
		-culling_limit="${CULLING}" \
		-outfmt "6 qseqid sseqid sacc pident length mismatch gapopen qcovus qstart qend sstart send evalue bitscore staxids qlen sscinames sseq" \
		-out "${blast_output}"
```






