---
title: "COI.denoise"
author: "M Fisher, via Eily Allen"
date: "2/9/2021"
output: 
  html_document:
    toc: yes
---

# Description 

This script takes output from cutadapt and dada2, and completes the following cleaning / denoising steps:

1. estimation of tag-jumping or cross-talk
2. discarding PCR replicates with a low number of reads
3. subtracting contamination in the controls from the samples
4. occupancy modeling

<br>

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
<br>

This script requires the following packages:
```{r message=FALSE}
 library (tidyverse)
 library(magrittr)
 library (vegan)
 library (here)
 library (proxy)
 library (reshape2)
 library (seqinr)
 library (patchwork)
```
<br>

Specify the markers.
```{r}
m <- "Leray"
```
<br>
<br>

## Data

We have completed cutadapt and dada2.

```{r load merged COI asv table, hash key, metadata}
metadata <- read.csv(here("test-run-1","data", "cutadapt", m, paste0("cutadapt_output_metadata_",m,".csv")))
ASV.table <- read.csv(here("test-run-1","data", "dada2", m, "merged_ASV_table.csv"))
Hash.key <- read.csv(here("test-run-1","data", "dada2", m, "merged_hash_key.csv"))
```
<br>

Clean up the metadata from `cutadapt`, and save it to the `dada2` folder.
```{r}
metadata %<>% dplyr::select(Sample_name, file1) %>%
  # from the file name, get the sample ID
  mutate(SampleID=str_remove(file1,"_L001_R1_001.fastq.fastq"))
if(m!="LerayXT"){
  metadata %<>% mutate(SampleID=str_remove(SampleID,paste0("Locus_",m,"_",m,"_")))
} else{
  metadata %<>% mutate(SampleID=str_remove(SampleID,paste0("Locus_",m,"_LXT_")))
}
metadata %<>% mutate(SampleID=ifelse(Sample_name >= 10, str_sub(SampleID,start=1, end=-5),
                                    str_sub(SampleID,start=1, end=-4))) %>%
  # from the sample ID, get the site, crab ID, and replicate (tech)
  mutate(site=ifelse(SampleID %in% c("Kangaroo","PCRNegative","SpeedVacNegative"), "Control", 
                     str_sub(SampleID, start=1,end=4))) %>%
  mutate(site=ifelse(site!="WASS",site,ifelse(grepl("2020",SampleID),"WASS 2020", "WASS 2018"))) %>%
  separate(col=SampleID, into=c("crab","tech"), sep=-1,remove=FALSE) %>%
  mutate(tech=tolower(tech)) %>%
  mutate(Type=ifelse(site=="Control","control","sample")) %>%
  # clean up control columns
  mutate(crab=ifelse(Type=="control", SampleID, crab),
         tech=ifelse(Type=="control","a",tech)) %>%
  rename(Sample.number=Sample_name) %>%
  dplyr::select(-file1)

if(m=="BF3"){
  # clean up the extra tech replicate I have for one of my crabs.
  metadata[which(metadata$SampleID=="WASS_2020_190a_25uL"),"crab"] <- "WASS_2020_190"
  metadata[which(metadata$SampleID=="WASS_2020_190a_25uL"),"tech"] <- "d"
}

write.csv(metadata,here("test-run-1","data","dada2_denoised",paste0(m,"_dada2_metadata.csv")),row.names=FALSE)

```
<br>
<br>


# 1: Estimation of *Tag-jumping* or sample *cross-talk*

Before we modify our datasets on any way, we can calculate how many sequences that were only supposed to be in the positives control appeared in the environmental samples, and how many did the opposite. First we divide the dataset into positive control and environmental samples. Also create an ordered list of the Hashes present in the positive controls, for ease of plotting

```{r split into two}
# first we need to rename the ASV table column from Sample_name to Sample.name
ASV.table %<>% rename(Sample.number=Sample_name)

ASV.table %<>% full_join(metadata,by="Sample.number")

ASV.table %>% 
  filter (Type == "control") %>% 
  group_by(Hash) %>% 
  summarise(tot = sum(nReads)) %>% 
  arrange(desc(tot)) %>% 
  pull(Hash) -> good.order
```

Now let's create a jumping vector. What proportion of the reads found in the positives control come from elsewhere, and what proportion of the reads in the samples come from the positives control.
<br>

### Step 1: Split the data into controls and samples

To streamline the process and make it easier to execute it similarly but independently on each Miseq run, we nest the dataset by run. 
So Step1 is create a nested table so we can run this analysis on each run independently. 


```{r nesting the dataset}
ASV.table %>% 
  group_by(Type) %>% 
  nest() %>% 
  pivot_wider(names_from = Type, values_from =  data) -> ASV.nested 
```

That wasn't too complicated. Let's start a summary function that keeps track of our cleaning process

```{r summary.file}
how.many <- function(ASV.table, round){
  ASV.table %>% ungroup() %>% 
    summarise(nsamples = n_distinct(SampleID),
              nHashes = n_distinct(Hash),
              nReads = sum(nReads), 
              Stage = paste0("Step_", round)) %>% 
    gather(starts_with("n"), value = "number", key = "Stat")
}

ASV.summary <- ASV.nested %>% 
  transmute(Summary = map(sample, ~ how.many(ASV.table = .,round = 0))) 
```
<br>

### Step 2: Model the composition of the positive controls of each run 

We create a vector of the composition of each positive control and substract it from the environmental samples from their runs
```{r jumping vector}
ASV.nested %<>% 
  mutate (contam.tibble = map(control, 
                              function(.x){
                                .x %>%
                                  group_by(SampleID) %>%
                                  mutate (TotalReadsperSample = sum(nReads)) %>%
                                  mutate (proportion = nReads/TotalReadsperSample) %>%
                                  group_by (Hash) %>%
                                  summarise (vector_contamination = max (proportion))
                                }) )
head(ASV.nested %>%
  select(contam.tibble) %>% 
  unnest(cols = contam.tibble))
```
<br>

### Step 3: Substract the composition of the positive controls from the environmental samples

The idea behind this procedure is that we know, for each run, how many reads from each Hash appeared in the positive controls. These come from 2 processes: sequences we know should appear in the positive controls, and sequences that have *jumped* from the environment to the positive controls. With this procedure, we subtract from every environmental sample the proportion of reads that jumped from elsewhere.

```{r cleaning step 1}
ASV.nested %<>% 
  mutate(cleaned.tibble = map2(sample, contam.tibble, function(.x,.y){ 
    .x %>%
      group_by (SampleID) %>%
      mutate (TotalReadsperSample = sum (nReads)) %>%
      left_join(.y, by = "Hash") %>%
      mutate (Updated_nReads = ifelse (!is.na(vector_contamination),  nReads - (ceiling(vector_contamination*TotalReadsperSample)), nReads)) %>%
      filter (Updated_nReads > 0) %>%
      ungroup() %>% 
      dplyr::select (SampleID, Hash, nReads = Updated_nReads)
        }))


head(ASV.nested %>% 
  select(cleaned.tibble) %>% 
  unnest(cleaned.tibble))
```
<br>

Add this step to the summary table we were creating
```{r summary.file.2}

ASV.summary %<>%
  bind_cols(ASV.nested %>% transmute( Summary.1 = map(cleaned.tibble, ~ how.many(ASV.table = .,round = "1.Jump"))))

```
<br>
<br>


# 2: Discarding PCR replicates with low number of reads

We will fit the number of reads assigned to each sample to a normal distribution and discard those samples with a probability of 95% of not fitting in that distribution. The output would be a dataset with less samples and potentially less number of unique Hashes.

```{r fitting nReads per sample, message=FALSE}
ASV.nested %>% 
  select(cleaned.tibble) %>% 
  unnest(cleaned.tibble) %>% 
  group_by(SampleID) %>%
  summarise(tot = sum(nReads)) -> all.reps

# Visualize
all.reps %>%  
  pull(tot) -> reads.per.sample

names(reads.per.sample) <- all.reps %>% pull(SampleID)  

normparams.reads <- MASS::fitdistr(reads.per.sample, "normal")$estimate

all.reps %>%  
  mutate(prob = pnorm(tot, normparams.reads[1], normparams.reads[2])) -> all.reps
```
```{r echo=FALSE, fig.width=4, fig.height=3}
ggplot(data=all.reps, aes(x=tot,y=prob)) +
  geom_point() +  
  geom_hline(aes(yintercept=0.075)) + geom_vline(aes(xintercept=normparams.reads[1])) +
  theme_bw()
```
<br>

Identify the outliers (lower left quadrant of graph above)
```{r}
outliers <- all.reps %>% 
  filter(prob < 0.075 & tot < normparams.reads[1])
```
<br>


```{r}

ASV.nested %<>% 
  mutate(Step.1.low.reads = map (cleaned.tibble, ~ filter(.,!SampleID %in% outliers$SampleID) %>% ungroup)) 

```
<br>

Add this step to the summary table we were creating
```{r}
ASV.summary %<>%
  bind_cols(ASV.nested %>% transmute( Summary.1 = map(Step.1.low.reads, ~ how.many(ASV.table = .,round = "2.Low.nReads"))))
```
<br>
<br>


# 3: Full clearance from Positive control influence

Removing the Hashes that belong to the positive controls. First, for each Hash that appeared in the positive controls, determine whether a sequence is a true positive or a true environment. For each Hash, we will calculate, maximum, mean and total number of reads in both positive and samples, and then we will use the following decission tree:

  * If all three statistics are higher in one of the groups, we will label it either of Environmental or Positive control influence.
  
  * If there are conflicting results, we will use the Hashes. to see if they belong to either the maximum abundance of a Hash is in a positive, then it is a positive, otherwise is a real sequence from the environment.

Now, for each Hash in each set of positives controls, calculate the proportion of reads that were missasigned - they appeared somewhere they were not expected.
We will divide that process in two: first . A second step would be to create a column named proportion switched, which states the proportion of reads from one Hash that jumped from the environment to a positive control or viceversa. The idea is that any presence below a threshold can be arguably belong to tag jumping.

```{r real or positive}
Hash.fate.step2 <- ASV.table %>% 
  rename(Sample.name=SampleID) %>%
  filter (Hash %in% good.order) %>%
  group_by(Sample.name) %>% 
  mutate(tot.reads = sum(nReads)) %>% 
  group_by(Hash,Sample.name) %>% 
  mutate(prop = nReads/tot.reads) %>% 
  group_by(Hash, Type) %>% 
  summarise (max.  = max(prop),
             mean. = mean(prop),
             tot.  = sum(nReads)) %>% 
  gather(contains("."), value = "number", key = "Stat") %>%
  spread(key = "Type", value = "number", fill = 0) %>% 
  group_by(Hash, Stat) %>%
  mutate(origin = case_when(control > sample ~ "Positive.control",
                            TRUE                ~ "Environment")) %>% 
  group_by (Hash) %>%
  mutate(tot = n_distinct(origin))

Hashes.to.remove.step2 <- Hash.fate.step2 %>% 
  filter(tot == 1) %>% 
  group_by(Hash) %>% 
  summarise(origin = unique(origin)) %>% 
  filter(origin == "Positive.control")

Hashes.to.remove.step2 %<>% full_join(ASV.table %>% 
                                        group_by(Type, Hash) %>% 
                                        summarise(ocurrences =n()) %>% 
                                        spread(key = Type, value = ocurrences, fill = 0) %>% 
                                        #left_join(Hashes.to.remove.step2) %>% 
                                        #mutate(origin = case_when(is.na(origin) ~ "Kept",
                                        #                         TRUE          ~ "Discarded")) %>% 
                                        mutate(second.origin = case_when(control >= sample ~ "Discarded",
                                                                         TRUE                 ~ "Kept")) %>% 
                                        filter(second.origin == "Discarded"))

## commented out what Ramon had here
# Hashes.to.remove.step2 %>% 
#   bind_rows(tibble(Hash = c("01204d587a3c3752f426f2f6d813c0ff2b80ec8b",
#                             "acebcd5c491bb273f3e4d615cafad649"))) -> Hashes.to.remove.step2
 
## i think we need to remove the technical replicates before this step... it is confusing the sample identifitaction
## for now, go back and make all of those samples not technical replicates and rerun the code 

```
IN order to train DADA2 to better distinguish when positive control sequences have arrived in the environment, we will keep the sequences in a csv file


```{r ASVs from positives}
Hashes.to.remove.step2 %>% 
  left_join(Hash.key) %>% 
  select(Hash, Sequence) %>% 
  write_csv(here("test-run-1","data", "dada2_denoised",paste0(m,".hashes.to.remove.csv")))

```
<br>

Remove the positive control hashes from the composition of the ASVs
```{r cleaning.Step2}
ASV.nested %<>% 
  mutate(Step2.tibble = map (Step.1.low.reads, ~ filter(.,!Hash %in% Hashes.to.remove.step2$Hash) %>% ungroup))

saveRDS(ASV.nested, file = here("test-run-1","data","dada2_denoised",paste0(m,".Cleaning.before.Occ.model.rds")))

ASV.nested <- readRDS(file = here("test-run-1","data","dada2_denoised",paste0(m,".Cleaning.before.Occ.model.rds")))
```
<br>

```{r}
ASV.summary %<>% bind_cols(ASV.nested %>% transmute( Summary.1 = map(Step2.tibble, ~ how.many(ASV.table = .,round = "3.Positives"))))
```
<br>
<br>


####################### ---- stopped here Feb 9 ---- #######################

# 4: Occupancy modeling

What is the probability of a true positive presence of a Hash in a Miseq Run. We will use eDNA occupancy modeling to asses whether a hash is a rare variant that spilled out or a true presence.

The process requires to load extra packages, create some model file, and group the hashes by Run, and biological replicate, summarizing the data in a presence absence format.

The occupancy model itself was performed in the Rmarkdown file `Rjags.tunning.Rmd`, so here we will upload the csv file that contains all probability of occurrences of all hashes per site. Each Hash-Site combination produces a matrix of presence absences that feeds the model - for some cases it is a 30x3 matrix, for others it is a 39x3. We summarized the number of occurrences in each case and run models for each unique case (to save computing time). Each unique model was run 10 times to filter out cases in which the model converge into a local maxima.

So we will import the object `Occ.fate.csv` and reduce the dataset to those Hashes with an occ > 0.8

```{r importing Occ results}
occ.results <- read_csv("../Input/Occ.fate.2020.csv")
occ.results %>% 
  ggplot(aes(x = max)) +
  geom_histogram(fill = "blue") +
  labs(x = "",
       y = "Number of Hashes")  -> left.plot
  
occ.results %>% 
  left_join(ASV.nested %>% 
              select(Step2.tibble) %>% 
              unnest(cols = Step2.tibble) %>% 
              group_by(Hash) %>% 
              summarise (tot = sum(nReads))) %>% 
  ggplot(aes(x = cut_interval(max, n = 20))) +
  geom_col(aes(y = tot), fill = "red", position = "stack") +
  
  labs(x = "",
       y = "Number of Reads")   -> right.plot
left.plot + right.plot 
```

So we will throw away most of the Hashes, but will keep most of the reads - we are getting into something here

```{r actual filtering}
 occ.results %>% 
  filter(max > 0.8) %>% 
  pull (Hash) -> to.keep
ASV.nested %>% 
  mutate(Step3.tibble = map (Step2.tibble, ~ filter(.,Hash %in% to.keep))) -> ASV.nested
ASV.nested %>% 
  transmute(Summary.1 = map(Step3.tibble, ~ how.many(ASVtable = .,round = "4.Occupancy"))) %>% 
  left_join(ASV.summary) %>% 
  mutate(Summary   = map2(Summary, Summary.1, bind_rows)) %>%
  dplyr::select(-Summary.1) -> ASV.summary 
```


# 5: Dissimilarity between PCR replicates

So, a second way of cleaning the dataset is to remove samples for which the dissimilarity between PCR replicates exceeds the normal distribution of dissimilarities.
Sometimes the preparation of a PCR replicate goes wrong for a number of reasons - that leads to a particular PCR replicate to be substantially different to the other 2. In that case, we will remove the PCR replicate that has higher dissimilarity with the other two.

The process starts by adding the biological information to the ASV table, then diving the dataset by their biological replicate. This will also remove any sample that is not included in the metadata, eg coming from a different project.

```{r dissimilarity between PCR replicates}
ASV.nested %>% 
  select(Miseq_run, Step3.tibble) %>% 
  unnest(Step3.tibble) %>%
  separate(sample, into = "original_sample", sep = "\\.", remove = F) -> cleaned.tibble
```


```{r quick check}
# do all samples have a name
cleaned.tibble %>% 
  filter (sample == "")
# do all of them have an original sample
cleaned.tibble %>% 
  filter(original_sample == "")
# do all of them have a Hash
cleaned.tibble %>% 
  filter(is.na(Hash))
# How many samples, how many Hashes
cleaned.tibble %>%
  ungroup %>% 
  summarise(n_distinct(sample), # 770
            n_distinct(Hash))   # 4571
# Let's check the levels of replication
cleaned.tibble %>% 
  group_by(original_sample) %>% 
  summarise(nrep = n_distinct(sample)) %>% 
  #filter (nrep == 2) # 13
  filter (nrep == 1) # 0 
```
Ok, so there are 13 samples for which we only have 2 PCR replicates1.   We will get rid of those with only 1, as we can't estimate the PCR bias there. THis is 

```{r remove single replicates}
discard.1 <- cleaned.tibble %>% 
  group_by(original_sample) %>% 
  mutate(nrep = n_distinct(sample)) %>% 
  #filter (nrep == 2) # 25
  filter (nrep == 1) %>% 
  distinct(sample) %>% pull(sample)
cleaned.tibble %>% 
  filter(!sample %in% discard.1) -> cleaned.tibble
```

Anyway, let's have a visual representation of the dissimilarities between PCR replicates, biological replicates and everything else.

```{r lets do the PCR replication}
cleaned.tibble %>%
  group_by (sample) %>%
  mutate (Tot = sum(nReads),
          Row.sums = nReads / Tot) %>% 
  group_by (Hash) %>%
  mutate (Colmax = max (Row.sums),
          Normalized.reads = Row.sums / Colmax) -> cleaned.tibble
tibble_to_matrix <- function (tb) {
  
  tb %>% 
    group_by(sample, Hash) %>% 
    summarise(nReads = sum(Normalized.reads)) %>% 
    spread ( key = "Hash", value = "nReads", fill = 0) -> matrix_1
    samples <- pull (matrix_1, sample)
    matrix_1 %>% 
      ungroup() %>% 
    dplyr::select ( - sample) -> matrix_1
    data.matrix(matrix_1) -> matrix_1
    dimnames(matrix_1)[[1]] <- samples
    vegdist(matrix_1) -> matrix_1
}
tibble_to_matrix (cleaned.tibble) -> all.distances.full
# Do all samples have a name?
summary(is.na(names(all.distances.full))) # Yes they do
```

Let's make the pairwise distances a long table
```{r}
as_tibble(subset(melt(as.matrix(all.distances.full)))) -> all.distances.melted
# Any mjor screw ups
summary(is.na(all.distances.melted$value))
# Now, create a three variables for all distances, they could be PCR replicates, BIOL replicates, or from the same site
all.distances.melted %>%
  separate (Var1, into = "Bottle1", sep = "\\.", remove = FALSE) %>%
  separate (Bottle1, into = "Site1", remove = FALSE) %>%
  separate (Var2, into ="Bottle2", sep = "\\.", remove = FALSE) %>%
  separate (Bottle2, into = "Site2", remove = FALSE) %>%
  mutate ( Day.site1 = str_sub(Bottle1, start = 1, end = -2),
           Day.site2 = str_sub(Bottle2, start = 1, end = -2),
           Distance.type = case_when( Bottle1 == Bottle2 ~ "PCR.replicates",
                                      Day.site1 == Day.site2 ~ "Biol.replicates",
                                      Site1 == Site2 ~ "Same Site",
                                      TRUE ~ "Different Site"
                                     )) %>%
  dplyr::select(Sample1 = Var1, Sample2 = Var2 , value , Distance.type) %>%
  filter (Sample1 != Sample2) -> all.distances.to.plot
# Checking all went well
sapply(all.distances.to.plot, function(x) summary(is.na(x))) # good boi
all.distances.to.plot$Distance.type <- all.distances.to.plot$Distance.type  %>% fct_relevel( "PCR.replicates", "Biol.replicates", "Same Site")
  ggplot (all.distances.to.plot ) +
  geom_histogram (aes (fill = Distance.type, x = value, after_stat(ndensity)), position = "dodge",  alpha = 0.9, bins = 50) +
  facet_wrap( ~ Distance.type) +
  labs (x = "Pairwise dissimilarity", y = "density" ,
        Distance.type = "Distance") +
    guides (fill = "none")
ggsave("visual.anova.png", dpi = "retina")
```

So our the distribution of dissimilarities is as we expected : lowest in technical replicates, then biological replicates and higher across our study system. Now let's see if there are any technical replicates that should be discarded due to their higher dissimilarity. We will calculate the distance from each PCR replicate to their group centroid, fit those distances to a normal distribution and discard values that are too high

```{r}
# Instead of chosing based on the pw distances, we can do a similar thing using the distance to centroid
# Find out which samples have only two pcr replicates
cleaned.tibble %>% dplyr::select(-Miseq_run) %>% group_by(original_sample) %>% nest() -> nested.cleaning
nested.cleaning %>% 
  mutate(matrix = map(data, tibble_to_matrix)) -> nested.cleaning
nested.cleaning %>% mutate(ncomparisons = map(matrix, length)) -> nested.cleaning
 
  
dist_to_centroid <- function (x,y) {
  biol <- rep(y, length(x))
  
  if (length(biol) == 1) {
    output = rep(x[1]/2,2)
    names(output) <- attr(x, "Labels")
  }else{ 
    
  dispersion <- betadisper(x, group = biol)
  output = dispersion$distances
  }
  output
    }
nested.cleaning <- nested.cleaning %>% mutate (distances = map2(matrix, original_sample, dist_to_centroid))
unlist (nested.cleaning$distances) -> all_distances
hist(all_distances)
```

```{r}
#normparams <- fitdistr(all_pairwise_distances, "normal")$estimate
normparams <- MASS::fitdistr(all_distances, "normal")$estimate                                      
#  probs <- pnorm(all_pairwise_distances, normparams[1], normparams[2])
probs <- pnorm(all_distances, normparams[1], normparams[2])
outliers <- which(probs>0.95)
discard <-names (all_distances[outliers])
to_write_discarded <- as.data.frame(all_distances[outliers]) %>% rownames_to_column("sample") %>% dplyr::select(sample, 
                                                                                                     distance_to_centroid = 2)
to_write_discarded <- to_write_discarded %>% bind_rows(tibble(sample = discard.1,
                                                              distance_to_centroid = NA))
write_csv(to_write_discarded ,"../Output/discared_samples.csv")
# Who passes this filter
all_distances %>%
  as.tibble() %>% 
  mutate(sample = names(all_distances)) %>% 
  filter(!sample %in% to_write_discarded$sample) %>% 
  separate(sample, into = "event", sep = -3) %>% 
  group_by(event) %>% 
  summarise(cases = n()) %>% 
  separate(event, into = c("Site", "Date"), remove = F) %>% 
  filter(Date != "201703") %>% 
  mutate(Date = lubridate::ymd(paste0(Date, "01"))) %>% 
  mutate(Area = case_when(Site %in% c("CP", "LK", "FH") ~ "San Juan Island",
                          TRUE                          ~ "Hood Canal"),
         Site = fct_recode(Site, "Cattle Point" = "CP",
                           "Lime Kiln" = "LK",
                           "Friday Harbor" = "FH",
                           "Twanoh" = "TW",
                           "Potlatch"= "PO",
                           "Lilliwaup" = "LL",
                           "Triton Cove"= "TR",
                           "Salsbury Park"= "SA")) -> Coverage.dataset 
Coverage.dataset %>% 
  ggplot()+
  geom_raster(aes(x= Date, y = Site, fill = cases))+
  geom_text(aes(x= Date, y = Site, label = cases)) +
  facet_wrap(~Area, ncol = 1, scales = "free_y")
  
  
  
  
```

Finally, let's remove these samples from the dataset

```{r actual cleaning}
ASV.nested %>% 
  mutate(Step4.tibble = map (Step3.tibble,  ~ filter(.,! sample %in% to_write_discarded$sample))) -> ASV.nested
ASV.nested %>% 
  transmute(Summary.1 = map(Step4.tibble, ~ how.many(ASVtable = .,round = "5.PCR.dissimilarity"))) %>% 
  left_join(ASV.summary) %>% 
  mutate(Summary   = map2(Summary, Summary.1, bind_rows)) %>%
  dplyr::select(-Summary.1) -> ASV.summary 
```


## Exporting the output

We will export the final cleaned table with four columns (Miseq_run, sample, Hash, nReads)

```{r}
ASV.nested %>% 
  select(Miseq_run, Step4.tibble) %>% 
  unnest(Step4.tibble) %>% 
  mutate(nReads = as.integer(nReads)) %>% 
  write_csv("../Output/ASV_table_all_together.csv")
ASV.nested %>% 
  select(Step4.tibble) %>% 
  unnest(Step4.tibble) %>% 
  distinct(Hash) %>% 
  left_join(Hash.key) %>% 
  write_csv("../Output/Hash_Key_all_together.csv")
input <- read_csv("../Output/Hash_Key_all_together.csv")
output <- "../Output/Hash_Key_all_together.fasta"
write.fasta (sequences = as.list(input$Sequence),
             names = as.list(input$Hash),
             file.out = output)
```

## Summary of the cleanup process

```{r last graph}
ASV.summary %>% 
  unnest() %>% 
  ggplot(aes(x=Stage, y=number, fill = Stat))+
    geom_line(aes(group = Miseq_run, color = Miseq_run))+
  facet_grid(Stat~., scales = "free")+
  theme(axis.text.x = element_text(angle = 45, hjust =1))#,
                                 
```

# Coverage Table with nReads

```{r}
# Vector with desired order
order.Sites <- c("Salsbury Park", "Triton Cove", "Lilliwaup", "Potlatch", "Twanoh", "Cattle Point", "Lime Kiln", "Friday Harbor")
ASV.nested %>% 
  select(Miseq_run, Step4.tibble) %>% 
  unnest(Step4.tibble) %>% 
  mutate(nReads = as.integer(nReads)) %>% 
  group_by(sample) %>% 
  summarise (nReads = sum(nReads)) %>% 
  separate (sample, into = "event", sep = -3) -> midstep
    
# As a table
midstep %>% 
  group_by(event) %>% 
  summarise (mean = mean(nReads), max = max(nReads), min = min(nReads), sd = sd(nReads)) %>% 
transmute (event, data = paste0("(", round(mean,0), " +- ", round(sd,0), ")")) %>% 
  right_join(Coverage.dataset) %>% 
  mutate( data = paste0(cases, " ", data)) %>% 
  arrange(Date) %>% 
  mutate (Date = paste(lubridate::month(Date, label = T),
                      lubridate::year(Date),
                      sep= "'")) %>% 
  pivot_wider(names_from = Date, values_from = data, id_cols = c(Area, Site),values_fill = list(data = "") ) %>% 
  slice(match(Site, order.Sites)) %>% 
  kableExtra::kable() %>% 
  kableExtra::kable_styling() %>% 
  kableExtra::collapse_rows(1, valign = "top") 
# As a graph
midstep %>% 
  right_join(Coverage.dataset) %>% 
  group_by(event) %>% 
  mutate(mean = mean(nReads),
         sum = sum(nReads)) %>% 
  mutate(Site = factor(Site, levels = order.Sites)) %>% 
  ggplot(aes(x = Date, 
             y = nReads)) +
  geom_boxplot(aes(group=Date),outlier.alpha = 0.5) +
  geom_point(aes(y = sum,
                 size = cases,
                 color = Area)) +
  scale_y_continuous(labels = scales::comma) +
  scale_x_date(labels = function(x){paste(lubridate::month(x, label = T),
                      lubridate::year(x),
                      sep= "'")}) +
  labs(y = "#reads", x = "") +
  
  facet_wrap(~Site , nrow = 5,dir = "v", drop = T) +
  theme_minimal() +
  theme(axis.title.y = element_text(angle = 0,vjust = 0.5),
        legend.position = c(0.6, 0.2)) 
```