---
title: "cutadapt.wrapper"
author: "Eily via Moncho"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: html_document
params:
  run.num:
    value: 6
  minlength:
    value: 100
---

## *IMPORTANT* won't work unless manually put in full file path to params.txt file at bottom - do that then knit.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(here)


# if this doesn't work (command not found), use the following:
old_path <- Sys.getenv("PATH")
Sys.setenv(PATH = paste(old_path, "/Users/elizabethandruszkiewicz/opt/anaconda2/bin/", sep = ":"))

setwd("/Users/elizabethandruszkiewicz/GoogleDrive/UW/GitHub/NextGenNEPA_EA/In_Progress")

# hard code in fastq folder to come from local hard drive NOT the github folder because the files are too big to host on Github
fastqfolder <- paste0("/Users/elizabethandruszkiewicz/GoogleDrive/UW/GitHub/NextGenNEPA_LOCAL/Input/fastq_files","/run",params$run.num)
sequencingmetadatafile <- paste0(here("Input","sequencing_metadata_files/"),"metadata-input-run",params$run.num, ".csv") 
outputfolder <- paste0("/Users/elizabethandruszkiewicz/GoogleDrive/UW/GitHub/NextGenNEPA_LOCAL/Output/","cutadapt_output","/run",params$run.num,"_",format(Sys.Date(), "%Y%m%d"))
paramsfile <- paste0(outputfolder,"/params.txt")

```

## Cutadapt wrapper

This is an R Markdown document. It is just a means to passing the parameters from the metadata file into a bash script

## Step0: Check the metadata colnames.

The minimum info needed must have the following column names:

        - Sample_name - The name that makes sense to you and your project (No spaces in the name would be better)
        - Locus: The name of the locus you want to use (e.g. Leray_COI)
        - PrimerF: The nucleotide sequence of the forward primer - supports IUPAC characters 
        - PrimerR: Ditto for the reverse primer (also in 5' -> 3' direction)
        - file1: it should match exactly the output of the Miseq.
        - file2: Same for the second read.
        
Does not matter if you bring more columns - good. But these are the minimum

## Step1: Params into a file for your records

```{r}
dir.create(outputfolder)

library(tidyverse)

params2 <- c(fastqfolder,sequencingmetadatafile, outputfolder, params$minlength)

tibble(values = as.character(params2), names = c("fastqfolder", "sequencingmetadatafile", "outputfolder", "minlength")) %>% 
  pivot_wider(names_from = names,
              values_from = values) %>%
  write_csv(paramsfile)


```

```{bash}
cutadapt --version
```


```{bash}

bash test.bash.sh /Users/elizabethandruszkiewicz/GoogleDrive/UW/GitHub/NextGenNEPA_LOCAL/Output/cutadapt_output/run6_20211229/params.txt
```
